frontend/src/contexts/AuthContext.tsx

```diff
@@ -0,0 +1,62 @@
+// contexts/AuthContext.tsx
+'use client';
+
+import { createContext, useContext, useEffect, useState, useCallback } from 'react';
+import { createClient } from '@/lib/supabase/client';
+import type { SupabaseClient, User, Session } from '@supabase/supabase-js';
+
+type AuthContextType = {
+  user: User | null;
+  session: Session | null;
+  isLoading: boolean;
+  isLoggedIn: boolean;
+  supabase: SupabaseClient;
+  refreshSession: () => Promise<void>;
+};
+
+const AuthContext = createContext<AuthContextType | undefined>(undefined);
+
+export function AuthProvider({ children }: { children: React.ReactNode }) {
+  const [user, setUser] = useState<User | null>(null);
+  const [session, setSession] = useState<Session | null>(null);
+  const [isLoading, setIsLoading] = useState(true);
+  const supabase = createClient();
+
+  const refreshSession = useCallback(async () => {
+    const { data } = await supabase.auth.getSession();
+    setSession(data.session);
+    setUser(data.session?.user || null);
+    setIsLoading(false);
+  }, [supabase]);
+
+  useEffect(() => {
+    refreshSession();
+
+    const { data: { subscription } } = supabase.auth.onAuthStateChange((_event, session) => {
+      setSession(session);
+      setUser(session?.user || null);
+      setIsLoading(false);
+    });
+
+    return () => subscription.unsubscribe();
```

Comment:
what does sub.unsubscribe do?

Reply:
the function returned by useEffect is the clean up function, so this will make sure we unsubscribe to the supabase auth listener when the component unmounts or is re-rendered